In [46]:
import pandas as pd
import numpy as np

df1 = "C:\\Users\\Dalit\\OneDrive\\Desktop\\Git_Hub_Repos\\Pandas_Challenge\\Resources\\schools_complete.csv"
df2 ="C:\\Users\\Dalit\\OneDrive\\Desktop\\Git_Hub_Repos\\Pandas_Challenge\\Resources\\students_complete.csv"




In [47]:
df1=pd.read_csv("C:\\Users\\Dalit\\OneDrive\\Desktop\\Git_Hub_Repos\\Pandas_Challenge\\Resources\\schools_complete.csv")
df2=pd.read_csv("C:\\Users\\Dalit\\OneDrive\\Desktop\\Git_Hub_Repos\\Pandas_Challenge\\Resources\\students_complete.csv")

df_merge= df1.merge(df2, on = "school_name", how = "inner")


In [48]:
df_merge

School ID         school_name      type  size   budget  Student ID  \
0              0   Huang High School  District  2917  1910635           0   
1              0   Huang High School  District  2917  1910635           1   
2              0   Huang High School  District  2917  1910635           2   
3              0   Huang High School  District  2917  1910635           3   
4              0   Huang High School  District  2917  1910635           4   
...          ...                 ...       ...   ...      ...         ...   
39165         14  Thomas High School   Charter  1635  1043130       39165   
39166         14  Thomas High School   Charter  1635  1043130       39166   
39167         14  Thomas High School   Charter  1635  1043130       39167   
39168         14  Thomas High School   Charter  1635  1043130       39168   
39169         14  Thomas High School   Charter  1635  1043130       39169   

            student_name gender grade  reading_score  math_score  
0           Paul Bradley      M   9th             66          79  
1           Victor Smith      M  12th             94          61  
2        Kevin Rodriguez      M  12th             90          60  
3      Dr. Richard Scott      M  12th             67          58  
4             Bonnie Ray      F   9th             97          84  
...                  ...    ...   ...            ...         ...  
39165       Donna Howard      F  12th             99          90  
39166          Dawn Bell      F  10th             95          70  
39167     Rebecca Tanner      F   9th             73          84  
39168       Desiree Kidd      F  10th             99          90  
39169    Carolyn Jackson      F  11th             95          75  

[39170 rows x 11 columns]

In [49]:
#Total Number of Students
named_students=df_merge["Student ID"].count()
named_students

39170

In [50]:
#Total Number of unique schools
named_schools=df_merge["School ID"].unique()
total_schools= len(named_schools)

total_schools

15

In [51]:
#Total Budget
total_school_budget=df_merge["budget"].sum()
total_school_budget

82932329558

In [52]:
# Average Math Score
av_math_score=df_merge["math_score"].mean()
av_math_score

78.98537145774827

In [53]:
#Average Reading score
av_reading_score=df_merge["reading_score"].mean()
av_reading_score

81.87784018381414

In [54]:
#Percentage of students who passed math  
passed_math = df_merge.loc[df_merge["math_score"] >= 70]
students_who_passed_math = passed_math["Student ID"].count()

passing_math_percentage = (students_who_passed_math /named_students) * 100

passing_math_percentage

74.9808526933878

In [55]:
#Percentage of students who passed reading
passed_reading = df_merge.loc[df2["reading_score"] >= 70]
students_who_passed_reading = passed_math["Student ID"].count()

passing_reading_percentage = (students_who_passed_reading /named_students) * 100

passing_reading_percentage

74.9808526933878

In [56]:
#budget for each student
per_student_budget = df_merge.set_index(["student_name"])["budget"]
per_student_budget

student_name
Paul Bradley         1910635
Victor Smith         1910635
Kevin Rodriguez      1910635
Dr. Richard Scott    1910635
Bonnie Ray           1910635
                      ...   
Donna Howard         1043130
Dawn Bell            1043130
Rebecca Tanner       1043130
Desiree Kidd         1043130
Carolyn Jackson      1043130
Name: budget, Length: 39170, dtype: int64

In [57]:
#Percentage of students who passed overall
overall_passing_percentage = df_merge[(df_merge["math_score"] >=70) & (df_merge["reading_score"] >=70)].groupby('Student ID')
                                      
overall_passing_percentage                            

In [58]:
overall_passing = (passing_math_percentage + passing_reading_percentage ) / 2
overall_passing

74.9808526933878

In [59]:
#district summary

district_results = [{"Total Schools": total_schools, 
            "Total Students": named_students, 
            "Total Budget": total_school_budget, 
            "Average Math Score":  av_math_score, 
            "Average Reading Score":  av_reading_score, 
           "% Passing Math": passing_math_percentage,
           "% Passing Reading":passing_reading_percentage,
            "% Overall Passing Rate": overall_passing_percentage,}]

district_summary= pd.DataFrame(district_results, columns = ["Total Schools",
                                                            "Total Students",
                                                            "Total Budget",
                                                            "Average Math Score",
                                                            "Average Reading Score",
                                                            "% Passing Math",
                                                            "% Passing Reading",
                                                            "% Overall Passing Rate"])
district_summary["Total Students"] =district_summary["Total Students"].map('{:,}'.format)
district_summary["Total Budget"] =district_summary["Total Budget"].map('${:,.2f}'.format)

                                                            

district_summary

Total Schools Total Students        Total Budget  Average Math Score  \
0             15         39,170  $82,932,329,558.00           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  \
0               81.87784       74.980853          74.980853   

                              % Overall Passing Rate  
0  <pandas.core.groupby.generic.DataFrameGroupBy ...

In [60]:
#SCHOOL SUMMARY
df_merge["passing_math_percentage"] = df_merge["math_score"] >= 70
df_merge["passing_reading_percentage"] = df_merge["reading_score"] >= 70

df_merge

School ID         school_name      type  size   budget  Student ID  \
0              0   Huang High School  District  2917  1910635           0   
1              0   Huang High School  District  2917  1910635           1   
2              0   Huang High School  District  2917  1910635           2   
3              0   Huang High School  District  2917  1910635           3   
4              0   Huang High School  District  2917  1910635           4   
...          ...                 ...       ...   ...      ...         ...   
39165         14  Thomas High School   Charter  1635  1043130       39165   
39166         14  Thomas High School   Charter  1635  1043130       39166   
39167         14  Thomas High School   Charter  1635  1043130       39167   
39168         14  Thomas High School   Charter  1635  1043130       39168   
39169         14  Thomas High School   Charter  1635  1043130       39169   

            student_name gender grade  reading_score  math_score  \
0           Paul Bradley      M   9th             66          79   
1           Victor Smith      M  12th             94          61   
2        Kevin Rodriguez      M  12th             90          60   
3      Dr. Richard Scott      M  12th             67          58   
4             Bonnie Ray      F   9th             97          84   
...                  ...    ...   ...            ...         ...   
39165       Donna Howard      F  12th             99          90   
39166          Dawn Bell      F  10th             95          70   
39167     Rebecca Tanner      F   9th             73          84   
39168       Desiree Kidd      F  10th             99          90   
39169    Carolyn Jackson      F  11th             95          75   

       passing_math_percentage  passing_reading_percentage  
0                         True                       False  
1                        False                        True  
2                        False                        True  
3                        False                       False  
4                         True                        True  
...                        ...                         ...  
39165                     True                        True  
39166                     True                        True  
39167                     True                        True  
39168                     True                        True  
39169                     True                        True  

[39170 rows x 13 columns]

In [62]:
# School Types and name
dfschool_types = df_merge.set_index(["school_name"])["type"]
dfschool_types

school_name
Huang High School     District
Huang High School     District
Huang High School     District
Huang High School     District
Huang High School     District
                        ...   
Thomas High School     Charter
Thomas High School     Charter
Thomas High School     Charter
Thomas High School     Charter
Thomas High School     Charter
Name: type, Length: 39170, dtype: object

In [63]:
#school summary 
total_stu_sch = df2.groupby('school_name')['student_name'].count()
total_bud_sch = df1.groupby('school_name')['budget'].sum()

av_math_scr_sch = df2.groupby('school_name')['math_score'].sum() / total_stu_sch
av_rea_scr_sch = df2.groupby('school_name')['reading_score'].sum() / total_stu_sch

pass_math_sch = df2.loc[df2['math_score'] > 70, :].groupby('school_name')['math_score'].count()
pass_rea_sch =df2.loc[df2['reading_score'] > 70, :].groupby('school_name')['reading_score'].count()

schools_df_all = pd.concat([total_stu_sch, 
                                 total_bud_sch, 
                                 av_math_scr_sch, 
                                 av_rea_scr_sch, 
                                 pass_math_sch, 
                                 pass_rea_sch], 
                               axis = 1)

schools_df_all.reset_index(inplace=True)

schools_df_all = schools_df_all.rename(columns={'index' : 'student_name',  
                                                       'student_name' : 'Total Students',
                                                        'budget' : 'Total School Budget',
                                                        0 : 'Average Math Score',
                                                        1 : 'Average Reading Score',
                                                        'math_score' : 'passing math new',
                                                        'reading_score' : 'passing reading new'
                                                        })

schools_df_all = df1.merge(schools_df_all, on='school_name')

schools_df_all['Per Student Budget'] = schools_df_all['Total School Budget'] / schools_df_all['Total Students']
schools_df_all['% Passing Math'] = (schools_df_all['passing math new'] / schools_df_all['Total Students']) * 100
schools_df_all['% Passing Reading'] = (schools_df_all['passing reading new'] / schools_df_all['Total Students']) * 100
schools_df_all['% Overall Passing Rate'] = (schools_df_all['% Passing Math'] + schools_df_all['% Passing Reading']) / 2

schools_summary = schools_df_all[['school_name', 
                                       'type', 
                                       'Total Students', 
                                       'Total School Budget', 
                                       'Per Student Budget', 
                                       'Average Math Score', 
                                       'Average Reading Score', 
                                       '% Passing Math', 
                                       '% Passing Reading', 
                                       '% Overall Passing Rate']]
                                       
schools_summary = schools_summary.rename(columns={'type' : 'School Type'})
schools_summary.set_index('school_name', inplace=True)
schools_summary.index.name

schools_summary['Total School Budget'] = schools_summary['Total School Budget'].map('$ {:,.2f}'.format)
schools_summary['Per Student Budget'] = schools_summary['Per Student Budget'].map('$ {:,.2f}'.format)


schools_summary


School Type  Total Students Total School Budget  \
school_name                                                             
Huang High School        District            2917      $ 1,910,635.00   
Figueroa High School     District            2949      $ 1,884,411.00   
Shelton High School       Charter            1761      $ 1,056,600.00   
Hernandez High School    District            4635      $ 3,022,020.00   
Griffin High School       Charter            1468        $ 917,500.00   
Wilson High School        Charter            2283      $ 1,319,574.00   
Cabrera High School       Charter            1858      $ 1,081,356.00   
Bailey High School       District            4976      $ 3,124,928.00   
Holden High School        Charter             427        $ 248,087.00   
Pena High School          Charter             962        $ 585,858.00   
Wright High School        Charter            1800      $ 1,049,400.00   
Rodriguez High School    District            3999      $ 2,547,363.00   
Johnson High School      District            4761      $ 3,094,650.00   
Ford High School         District            2739      $ 1,763,916.00   
Thomas High School        Charter            1635      $ 1,043,130.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Huang High School               $ 655.00           76.629414   
Figueroa High School            $ 639.00           76.711767   
Shelton High School             $ 600.00           83.359455   
Hernandez High School           $ 652.00           77.289752   
Griffin High School             $ 625.00           83.351499   
Wilson High School              $ 578.00           83.274201   
Cabrera High School             $ 582.00           83.061895   
Bailey High School              $ 628.00           77.048432   
Holden High School              $ 581.00           83.803279   
Pena High School                $ 609.00           83.839917   
Wright High School              $ 583.00           83.682222   
Rodriguez High School           $ 637.00           76.842711   
Johnson High School             $ 650.00           77.072464   
Ford High School                $ 644.00           77.102592   
Thomas High School              $ 638.00           83.418349   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Huang High School                  81.182722       63.318478   
Figueroa High School               81.158020       63.750424   
Shelton High School                83.725724       89.892107   
Hernandez High School              80.934412       64.746494   
Griffin High School                83.816757       89.713896   
Wilson High School                 83.989488       90.932983   
Cabrera High School                83.975780       89.558665   
Bailey High School                 81.033963       64.630225   
Holden High School                 83.814988       90.632319   
Pena High School                   84.044699       91.683992   
Wright High School                 83.955000       90.277778   
Rodriguez High School              80.744686       64.066017   
Johnson High School                80.966394       63.852132   
Ford High School                   80.746258       65.753925   
Thomas High School                 83.848930       90.214067   

                       % Passing Reading  % Overall Passing Rate  
school_name                                                       
Huang High School              78.813850               71.066164  
Figueroa High School           78.433367               71.091896  
Shelton High School            92.617831               91.254969  
Hernandez High School          78.187702               71.467098  
Griffin High School            93.392371               91.553134  
Wilson High School             93.254490               92.093736  
Cabrera High School            93.864370               91.711518  

In [64]:
#  average math score by grade 9-12th grade from each school


nineth_maths = df_merge.loc[df_merge['grade'] == '9th'].groupby('school_name')["passing_math_percentage"].mean()
tenth_maths = df_merge.loc[df_merge['grade'] == '10th'].groupby('school_name')["passing_math_percentage"].mean()
eleventh_maths = df_merge.loc[df_merge['grade'] == '11th'].groupby('school_name')["passing_math_percentage"].mean()
twelfth_maths= df_merge.loc[df_merge['grade'] == '12th'].groupby('school_name')["passing_math_percentage"].mean()


#merges the reading score averages by school and grade together
reading_scores = pd.DataFrame({
        "9th": nineth_maths,
        "10th": tenth_maths,
        "11th": eleventh_maths,
        "12th": twelfth_maths
})
maths_scores = reading_scores[['9th', '10th', '11th', '12th']]
maths_scores.index.name = "School Name",

#format
maths_scores.style.format({'9th': '{:.6f}', 
                             "10th": '{:.6f}', 
                             "11th": "{:.6f}", 
                             "12th": "{:.6f}"})

In [97]:

#  average reading score by grade 9-12th grade from each school

nineth_reading = df_merge.loc[df_merge['grade'] == '9th'].groupby('school_name')["passing_reading_percentage"].mean()
tenth_reading = df_merge.loc[df_merge['grade'] == '10th'].groupby('school_name')["passing_reading_percentage"].mean()
eleventh_reading = df_merge.loc[df_merge['grade'] == '11th'].groupby('school_name')["passing_reading_percentage"].mean()
twelfth_reading = df_merge.loc[df_merge['grade'] == '12th'].groupby('school_name')["passing_reading_percentage"].mean()


#merges the reading score averages by school and grade together
reading_scores = pd.DataFrame({
        "9th": nineth_reading,
        "10th": tenth_reading,
        "11th": eleventh_reading,
        "12th": twelfth_reading
})
reading_scores = reading_scores[['9th', '10th', '11th', '12th']]
reading_scores.index.name = "School Name"

#format
reading_scores.style.format({'9th': '{:0.2f}', 
                             "10th": '{:2f}', 
                             "11th": "{:.2f}", 
                             "12th": "{:.2f}"})

In [22]:
#defining variables for binning
spending_bins = [0, 585, 630, 645, 680]
#spending rsnges
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

In [95]:
#defining variables for binning
spending_bins = [0, 585, 630, 645, 680]
#spending ranges
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

# df columns
spending_amount = schools_summary.loc[:,['Total School Budget',
                                  'Per Student Budget','% Passing Math',
                                  '% Passing Reading','% Overall Passing Rate',]]
# Add a new columns named School Size and binning based off total students
spending_amount['budget']= pd.cut(schools_summary['Total Students'],spending_bins)
# Create a group based off of the bins
spending_amount = spending_amount.groupby('Per Student Budget').mean()
spending_amount.head()

% Passing Math  % Passing Reading  % Overall Passing Rate
Per Student Budget                                                           
$ 578.00                 90.932983          93.254490               92.093736
$ 581.00                 90.632319          92.740047               91.686183
$ 582.00                 89.558665          93.864370               91.711518
$ 583.00                 90.277778          93.444444               91.861111
$ 600.00                 89.892107          92.617831               91.254969

#Report

##Background


You are the new Chief Data Scientist for your city's school district. In this capacity, you'll be helping the school board and mayor make strategic decisions regarding future school budgets and priorities.

As a first task, you've been asked to analyze the district-wide standardized test results. You'll be given access to every student's math and reading scores, as well as various information on the schools they attend. Your task is to aggregate the data to showcase obvious trends in school performance.

## What Does the Above Data Mean  ?

As shown in the schools_summary table in cell 63, we can see Griffin High School is a Charter school with a budget of $625.00 per student and a pass rate of over 90%

In comparison, we see Huang High School a District school with a higher budget of $655 per student and an overall pass rate of just over 70%.

This trend remains similar when you closely examine all the results. For this reason we can safely conclude the following:
1 - Allocating a higher 'Per Student Budget' does not guarantee a higher pass rate outcome.
2- Students attending a Charter school perfomed well in their tests and this was irrespective of how low the 'Per Student Budget' was. 

Advise would be to compare the teaching methods at both schools and family backgroud and see if that has an impact on performance in schools. This data would then help chart the appropriae course of action. 
